src/client/api/object.c

```diff
@@ -397,75 +429,138 @@ struct oit_args {
-	 * implementation only has one bucket (dkey).
+	/* A bucket is just an integer dkey for OI object.
 	 */
 	uint32_t		 oa_bucket;
-	int			 oa_nr;
+	uint32_t		 oa_want_nr;
+	uint32_t		 oa_listed_nr;
 };
 
+static void
+anchor2buckets(daos_anchor_t *anchor, uint32_t *cur_bucket)
+{
+	char *ptr = (char *)&anchor->da_buf[DAOS_ANCHOR_BUF_MAX - 4];
+
+	memcpy((char *)cur_bucket, ptr, 4);
+}
+
+static void
+buckets2anchor(daos_anchor_t *anchor, uint32_t cur_bucket)
+{
+	char *ptr = (char *)&anchor->da_buf[DAOS_ANCHOR_BUF_MAX - 4];
+
+	memcpy(ptr, (char *)&cur_bucket, 4);
+}
+
 static int
-oit_list_cb(tse_task_t *task, void *args)
+oit_max_bucket(daos_handle_t oh, uint32_t *max_bucket)
 {
-	struct oit_args	*oa = *(struct oit_args **)args;
+	daos_handle_t	coh;
+	uint32_t	globalver;
+	int		rc;
+
+	coh = dc_obj_hdl2cont_hdl(oh);
+	if (!daos_handle_is_valid(coh))
+		return -DER_INVAL;
+
+	rc = dc_cont_hdl2globalver(coh, &globalver);
+	if (rc)
+		return rc;
+
+	if (globalver < 2)
+		*max_bucket = 1;
+	else
+		*max_bucket = DAOS_OIT_BUCKET_MAX;
 
-	d_sgl_fini(&oa->oa_sgl, false);
-	D_FREE(oa->oa_kds);
-	D_FREE(oa);
 	return 0;
 }
 
 int
 daos_oit_list(daos_handle_t oh, daos_obj_id_t *oids, uint32_t *oids_nr,
 	      daos_anchor_t *anchor, daos_event_t *ev)
 {
-	struct oit_args	*oa;
-	tse_task_t	*task;
 	int		 i;
-	int		 rc;
+	int		 rc = 0;
+	uint32_t	 max_bucket;
+	uint32_t	 bucket;
+	struct oit_args	*oa;
 
 	if (daos_handle_is_inval(oh) ||
 	    oids == NULL || oids_nr == NULL || *oids_nr <= 0 || !anchor)
 		return -DER_INVAL;
 
+	rc = oit_max_bucket(oh, &max_bucket);
+	if (rc)
+		return rc;
+
+	anchor2buckets(anchor, &bucket);
+	if (bucket >= max_bucket) {
+		*oids_nr = 0;
+		if (ev)
+			daos_event_complete(ev, 0);
+		return 0;
+	}
+
 	D_ALLOC_PTR(oa);
 	if (!oa)
 		return -DER_NOMEM;
 
-	oa->oa_nr = *oids_nr;
-	D_ALLOC_ARRAY(oa->oa_kds, oa->oa_nr);
+	oa->oa_want_nr = *oids_nr;
+	oa->oa_listed_nr = 0;
+	oa->oa_bucket = bucket;
+	D_ALLOC_ARRAY(oa->oa_kds, oa->oa_want_nr);
 	if (!oa->oa_kds)
 		D_GOTO(failed, rc = -DER_NOMEM);
 
-	rc = d_sgl_init(&oa->oa_sgl, oa->oa_nr);
+	rc = d_sgl_init(&oa->oa_sgl, oa->oa_want_nr);
 	if (rc)
 		D_GOTO(failed, rc = -DER_NOMEM);
 
-	for (i = 0; i < oa->oa_nr; i++)
-		d_iov_set(&oa->oa_sgl.sg_iovs[i], &oids[i], sizeof(oids[i]));
+	anchor2buckets(anchor, &oa->oa_bucket);
+	DAOS_OIT_DKEY_SET(&oa->oa_dkey, &oa->oa_bucket);
+	while (1) {
+		for (i = 0; i < *oids_nr; i++)
+			d_iov_set(&oa->oa_sgl.sg_iovs[i], &oids[i + oa->oa_listed_nr],
+				  sizeof(oids[i + oa->oa_listed_nr]));
 
-	/* all OIDs are stored under one dkey for now */
-	d_iov_set(&oa->oa_dkey, &oa->oa_bucket, sizeof(oa->oa_bucket));
-	rc = dc_obj_list_akey_task_create(oh, DAOS_TX_NONE, &oa->oa_dkey,
-					  oids_nr, oa->oa_kds, &oa->oa_sgl,
-					  anchor, ev, NULL, &task);
-	if (rc)
-		D_GOTO(failed, rc);
+		rc = daos_obj_list_akey(oh, DAOS_TX_NONE, &oa->oa_dkey, oids_nr,
+					oa->oa_kds, &oa->oa_sgl, anchor, NULL);
```

Comment:
here you turned asynchronous API calling "daos_oit_list(..., ev)" to a set of blocking synchronous API calls, and fake an event completion when exit the func, this looks not a good manner as it is a public daos async API.
To keep it to be async mode, need to be based on task and completion callback driven.

Reply:
in that case we should add a comment in the API header to mention async support is not supported at this time & complete the event at the end.
I can work on creating the sub-tasks for this later (but please create a ticket so i don't forget about it).

Reply:
I leave it as Mohamad suggestion for now.

Reply:
> in that case we should add a comment in the API header to mention async support is not supported at this time & complete the event at the end. I can work on creating the sub-tasks for this later (but please create a ticket so i don't forget about it).

I created https://daosio.atlassian.net/browse/DAOS-13159 for this.